File: src/TranslateTheDamn.Core/Util/AnsiStripper.cs
using System.Text;

namespace TranslateTheDamn.Core.Util;

/// <summary>
/// Removes ANSI escape sequences (CSI cursor/SGR colour codes + OSC) and carriage returns from
/// captured CLI output. Implemented as a manual scanner to avoid regex-escaping pitfalls.
/// </summary>
public static class AnsiStripper
{
    private const char Esc = (char)0x1B;   // ESC
    private const char Bel = (char)0x07;   // BEL

    public static string Strip(string? s)
    {
        if (string.IsNullOrEmpty(s)) return string.Empty;

        var sb = new StringBuilder(s.Length);
        int i = 0;
        while (i < s.Length)
        {
            char c = s[i];

            if (c == Esc && i + 1 < s.Length)
            {
                char next = s[i + 1];
                if (next == '[')
                {
                    // CSI: ESC '[' params... final-byte in 0x40..0x7E
                    i += 2;
                    while (i < s.Length && (s[i] < '@' || s[i] > '~')) i++;
                    if (i < s.Length) i++; // consume final byte
                    continue;
                }
                if (next == ']')
                {
                    // OSC: ESC ']' ... terminated by BEL or by ST (ESC '\')
                    i += 2;
                    while (i < s.Length && s[i] != Bel && !(s[i] == Esc && i + 1 < s.Length && s[i + 1] == '\\')) i++;
                    if (i < s.Length && s[i] == Bel) i++;
                    else if (i + 1 < s.Length) i += 2; // skip ESC '\'
                    continue;
                }
                // Any other escape: drop ESC + the following byte.
                i += 2;
                continue;
            }

            if (c == '\r') { i++; continue; }

            sb.Append(c);
            i++;
        }

        return sb.ToString();
    }
}

File: platforms/macos/src/Core/AnsiStripper.swift
import Foundation

/// Removes ANSI escape sequences (CSI cursor/SGR colour + OSC) and carriage returns from captured
/// CLI output (spec §6 — CLIs are tamed into clean text→text calls).
public enum AnsiStripper {
    public static func strip(_ s: String) -> String {
        var result = ""
        result.reserveCapacity(s.count)

        var i = s.startIndex
        let end = s.endIndex

        while i < end {
            let c = s[i]

            if c == "\u{1B}" {
                let next = s.index(after: i)
                guard next < end else {
                    // Lone ESC at end; drop it.
                    break
                }
                let nc = s[next]

                if nc == "[" {
                    // CSI: ESC '[' params/intermediates... final byte 0x40–0x7E
                    i = s.index(after: next)
                    while i < end {
                        let b = s[i]
                        guard let scalar = b.unicodeScalars.first,
                              scalar.value < 0x40 || scalar.value > 0x7E else {
                            break
                        }
                        i = s.index(after: i)
                    }
                    if i < end {
                        i = s.index(after: i) // consume final byte
                    }
                    continue
                }

                if nc == "]" {
                    // OSC: ESC ']' ... terminated by BEL or ST (ESC '\')
                    i = s.index(after: next)
                    while i < end {
                        let b = s[i]
                        if b == "\u{07}" {
                            i = s.index(after: i)
                            break
                        }
                        if b == "\u{1B}" {
                            let afterB = s.index(after: i)
                            if afterB < end && s[afterB] == "\\" {
                                i = s.index(after: afterB)
                                break
                            }
                        }
                        i = s.index(after: i)
                    }
                    continue
                }

                // Any other ESC sequence: drop ESC + the following byte.
                i = s.index(after: next)
                continue
            }

            if c == "\r" {
                i = s.index(after: i)
                continue
            }

            result.append(c)
            i = s.index(after: i)
        }

        return result
    }
}
